source/core_mqtt.c

```diff
@@ -70,6 +70,101 @@
+ * unsubscribe request namely:
+ * 1. Topic filter length; and 2. Topic filter in this order.
+ */
+#define CORE_MQTT_UNSUBSCRIBE_PER_TOPIC_VECTOR_LENGTH     ( 2U )
+
+/**
+ * @brief Bytes required to encode the packet ID in an MQTT header according
+ * to the MQTT specification.
+ */
+#define CORE_MQTT_SERIALIZED_PACKET_ID_BYTES              ( 2U )
+
+/**
+ * @brief Maximum number of bytes required by the 'fixed' part of the CONNECT
+ * packet header according to the MQTT specification.
+ * MQTT Control Byte      0 + 1 = 1
+ * Remaining length (max)   + 4 = 5
+ * Protocol Name Length     + 2 = 7
+ * Protocol Name (MQTT)     + 4 = 11
+ * Protocol level           + 1 = 12
+ * Connect flags            + 1 = 13
+ * Keep alive               + 2 = 15
+ */
+#define CORE_MQTT_CONNECT_PACKET_HEADER_MAX_LENGTH        ( 15U )
+
+/**
+ * @brief Maximum number of bytes required by the 'fixed' part of the SUBSCRIBE
+ * packet header according to the MQTT specification.
+ * MQTT Control Byte      0 + 1 = 1
+ * Remaining length (max)   + 4 = 5
+ * Packet ID                + 2 = 7
+ */
+#define CORE_MQTT_SUBSCRIBE_PACKET_HEADER_MAX_LENGTH      ( 7U )
+
+/**
+ * @brief Maximum number of bytes required by the 'fixed' part of the UNSUBSCRIBE
+ * packet header according to the MQTT specification.
+ * MQTT Control Byte      0 + 1 = 1
+ * Remaining length (max)   + 4 = 5
+ * Packet ID                + 2 = 7
+ */
+#define CORE_MQTT_UNSUBSCRIBE_PACKET_HEADER_MAX_LENGTH    ( 7U )
+
+/**
+ * @brief Maximum number of bytes required by the 'fixed' part of the PUBLISH
+ * packet header according to the MQTT specifications.
+ * Header byte           0 + 1 = 1
+ * Length (max)            + 4 = 5
+ * Topic string length     + 2 = 7
+ *
+ * @note Since publish is one of the most common operations in MQTT connection,
+ * we have moved the topic string length to the 'fixed' part of the header so
+ * efficiency. Otherwise, we would need an extra vector and an extra call to
+ * 'send' (in case writev is not defined) to send the topic length.
+ */
+#define CORE_MQTT_PUBLISH_PACKET_HEADER_MAX_LENGTH        ( 7U )
+
+/**
+ * @brief The maximum vectors required to encode and send a connect packet. The
+ * breakdown is shown below.
+ * Fixed header      0 + 1 = 1
+ * Client ID           + 2 = 3
+ * Will topic          + 2 = 5
+ * Will payload        + 2 = 7
+ * Username            + 2 = 9
+ * Password            + 2 = 11
+ */
+#define CORE_MQTT_CONNECT_PACKET_MAX_VECTORS              ( 11U )
+
+/**
+ * @brief Maximum number of vectors required to encode and send a publish
+ * packet. The breakdown is shown below.
+ * Fixed header (including topic string length)      0 + 1 = 1
+ * Topic string                                        + 1 = 2
+ * Packet ID (only when QoS > QoS0)                    + 1 = 3
+ * Payload                                             + 1 = 4
+ */
+#define CORE_MQTT_PUBLISH_PACKET_MAX_VECTORS              ( 4U )
```

Comment:
CORE_MQTT_PUBLISH_PACKET_MAX_VECTORS is used in only one place. Moving this comment down there could improve clarity.